server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -73,6 +78,27 @@ private void checkForDuplicateAliases(Collection<SystemIndexDescriptor> descript
         }
     }
 
+    private static Map<String, CharacterRunAutomaton> getProductToSystemIndicesMap(Collection<SystemIndexDescriptor> descriptors) {
+        Map<String, Automaton> map = new HashMap<>();
+        descriptors.stream().filter(SystemIndexDescriptor::isExternal)
+            .forEach(descriptor ->
+                descriptor.getAllowedElasticProductOrigins().forEach(product ->
+                    map.compute(product, (key, automaton) -> {
+                        Automaton built = SystemIndexDescriptor.buildAutomaton(descriptor.getIndexPattern(), descriptor.getAliasName());
+                        if (automaton != null) {
+                            return Operations.union(automaton, built);
+                        } else {
+                            return built;
+                        }
+                    })
+                )
+            );
```

Comment:
What do you think about using a `.flatMap()` here to flatten out the stream processing?

```suggestion
        descriptors.stream()
            .filter(SystemIndexDescriptor::isExternal)
            .flatMap(descriptor -> descriptor.getAllowedElasticProductOrigins().stream())
            .forEach(product ->
                map.compute(product, (key, automaton) -> {
                    Automaton built = SystemIndexDescriptor.buildAutomaton(
                        descriptor.getIndexPattern(), descriptor.getAliasName());
                    return automaton == null ? built : Operations.union(automaton, built);
                })
            );
```